barretenberg/cpp/src/barretenberg/commitment_schemes/gemini/gemini.cpp

```diff
@@ -1,237 +1,6 @@
-#include "gemini.hpp"
 #include "barretenberg/common/thread.hpp"
-
-/**
- * @brief Protocol for opening several multi-linear polynomials at the same point.
- *
- *
- * m = number of variables
- * n = 2ᵐ
- * u = (u₀,...,uₘ₋₁)
- * f₀, …, fₖ₋₁ = multilinear polynomials,
- * g₀, …, gₕ₋₁ = shifted multilinear polynomial,
- *  Each gⱼ is the left-shift of some f↺ᵢ, and gⱼ points to the same memory location as fᵢ.
- * v₀, …, vₖ₋₁, v↺₀, …, v↺ₕ₋₁ = multilinear evalutions  s.t. fⱼ(u) = vⱼ, and gⱼ(u) = f↺ⱼ(u) = v↺ⱼ
- *
- * We use a challenge ρ to create a random linear combination of all fⱼ,
- * and actually define A₀ = F + G↺, where
- *   F  = ∑ⱼ ρʲ fⱼ
- *   G  = ∑ⱼ ρᵏ⁺ʲ gⱼ,
- *   G↺ = is the shift of G
- * where fⱼ is normal, and gⱼ is shifted.
- * The evaluations are also batched, and
- *   v  = ∑ ρʲ⋅vⱼ + ∑ ρᵏ⁺ʲ⋅v↺ⱼ = F(u) + G↺(u)
- *
- * The prover then creates the folded polynomials A₀, ..., Aₘ₋₁,
- * and opens them at different points, as univariates.
- *
- * We open A₀ as univariate at r and -r.
- * Since A₀ = F + G↺, but the verifier only has commitments to the gⱼs,
- * we need to partially evaluate A₀ at both evaluation points.
- * As univariate, we have
- *  A₀(X) = F(X) + G↺(X) = F(X) + G(X)/X
- * So we define
- *  - A₀₊(X) = F(X) + G(X)/r
- *  - A₀₋(X) = F(X) − G(X)/r
- * So that A₀₊(r) = A₀(r) and A₀₋(-r) = A₀(-r).
- * The verifier is able to computed the simulated commitments to A₀₊(X) and A₀₋(X)
- * since they are linear-combinations of the commitments [fⱼ] and [gⱼ].
- */
+#include "gemini_impl.hpp"
```

Comment:
We need to template by Transcript to make things work with both Ultra and UltraKeccak (plus we might want to use a test transcript). Hence, moved the functions from source file to an impl file to avoid ugly explicit instantiations.